InnerEye/ML/SSL/lightning_modules/ssl_online_evaluator.py

```diff
@@ -57,22 +57,35 @@ def __init__(self,
                                           lr=self.learning_rate,
                                           weight_decay=self.weight_decay)
 
+    def _wrapped_evaluator(self) -> torch.nn.Module:
+        """
+        Gets the evaluator model that is wrapped in DDP, or the evaluator model itself.
+        """
+        if isinstance(self.evaluator, DistributedDataParallel):
+            return self.evaluator.module
+        else:
+            return self.evaluator
+
     def on_save_checkpoint(self,
                            trainer: pl.Trainer,
                            pl_module: pl.LightningModule,
                            checkpoint: Dict[str, Any]) -> Dict[str, Any]:
         # Each callback gets its own state dictionary, that are fed back in during load
+        # When saving the evaluator, use the wrapped DDP module (otherwise the resulting checkpoint will depend
+        # on use of DDP or not).
         return {
             self.OPTIMIZER_STATE_NAME: self.optimizer.state_dict(),
-            self.EVALUATOR_STATE_NAME: self.evaluator.state_dict()
+            self.EVALUATOR_STATE_NAME: self._wrapped_evaluator().state_dict()
         }
 
     def on_load_checkpoint(self,
                            trainer: pl.Trainer,
                            pl_module: pl.LightningModule,
                            callback_state: Dict[str, Any]) -> None:
         self.optimizer.load_state_dict(callback_state[self.OPTIMIZER_STATE_NAME])
-        self.evaluator.load_state_dict(callback_state[self.EVALUATOR_STATE_NAME])
+        # on_load_checkpoint is called before on_pretrain_routine_start, where we wrap the evaluator with DDP.
```

Comment:
Isn't this the other way around (on_load_checkpoint is called after on_pretrain_routine_start)?

Reply:
No (and I just double-checked by running a test). First is on_load_checkpoint, then on_pretrain_routine_start